twoftpd-xfer.1
.TH twoftpd-xfer 1
.SH NAME
twoftpd-xfer \- TwoFTPD file transfer back-end.
.SH SYNOPSIS
.B twoftpd-xfer
.SH DESCRIPTION
.B twoftpd-xfer
reads FTP request from the network and executes file
transfers based on those requests.
Before doing anything,
.B twoftpd-xfer
changes directory to the authenticated user's home directory, changes
the root directory to that directory, and drops all root priviledges.
.SH ENVIRONMENT
.TP
.B AUTHENTICATED
If set (typically by
.B twoftpd-auth
after successfully completing authentication), the startup banner will
use code
.IR 230 ,
indicating that authentication was successful.  Otherwise, the startup
banner will use code
.IR 220 ,
which is the normal initial connection banner.
.TP
.B BANNER
If set, the contents of this variable will be sent as extra responses when
the server starts.
.TP
.B CHROOT
If set, change the root directory to the user's home directory through
the OS chroot syscall.
.TP
.B CONNECT_TIMEOUT
The maximum number of seconds to wait for a data channel connection to
complete.  If not set, it defaults to the value of
.BR $TIMEOUT .
.TP
.B GID
Numerical group ID number to change to at program startup.
.TP
.B GIDS
Optional comma-seperated list of numerical group IDs to set up as the
supplementary group IDs.
.TP
.B GROUP
Group name to use in long listings for files owned by
.BR $GID .
If not set, defaults to
.IR mygroup .
.TP
.B HOME
Directory to switch to on program startup.
.TP
.B LOCKHOME
If set (and neither
.B CHROOT
nor
.B SOFTCHROOT
are set), all accesses to files and paths outside of
.B HOME
will be forbidden.
.TP
.B LOGREQUESTS
If set, all requests from the client will be logged.
.TP
.B LOGRESPONSES
If set, all responses to the client (after login) will be logged.
.TP
.B MESSAGEFILE
If set, the contents this file will be automatically sent as extra
responses to the client whenever the server enters a directory
(including at startup).
.TP
.B NODOTFILES
If set, all listings will exclude all files starting with a period,
and it will not be possible to do any file access (including storing
or retrieving) to any file starting with a period.  This may be useful
for anonymous FTP servers.
.TP
.B SESSION_TIMEOUT
The maximum number of seconds to allow a single session to last.
.TP
.B SOFTCHROOT
If set, and
.B CHROOT
is not set, the server emulates the effect of a chroot without
actually effecting a chroot.  This eliminates the system-enforced
barrier provided by the chroot, but allows for things like symlinks
outside of the home directory.
.TP
.B TCPLOCALIP
The IP to use when creating PASV connections.
Normally, this is set by
.BR tcpserver .
.TP
.B TCPREMOTEIP
The source IP to use when creating PORT connections.
Normally, this is set by
.BR tcpserver .
.TP
.B TIMEOUT
The maximum number of seconds to wait for command input or
connections.
.TP
.B UID
Numerical user ID number to change to at program startup.
.TP
.B USER
Owner name to use in long listings for files owned by
.BR $UID .
.SH NOTES
You will need to set the umask to an appropriate value before running
this program.
.P
The
.B twoftpd-anon
program functions identically to this program, with the exception of
having no file modification capabilities.
The
.B twoftpd-drop
program is a similarly restricted program, that can only create new
files (no overwriting, modifying, or listing existing files).


NEWS
-------------------------------------------------------------------------------
Changes in version 1.41

- Fix a one-off buffer overrun in ASCII conversion in the STOR command.

- Fix resetting the output buffer between listings.

- Added a $CONNECT_TIMEOUT setting to set timeouts on data connections.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.40

- File transfer handling has been completely rewritten to handle
  interrupts on the command channel.  Closing the command channel or
  sending another command will interrupt any data transfer in progress.

- Fixed a longstanding bug in the handling of $MESSAGEFILE

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.30

- Switched to the CVM v2 client library.

- Added a $AUTH_ATTEMPTS control to twoftpd-auth to limit the number of
  authentication attempts per session.

- Rewrote the directory listing code to fix listing files with special
  characters in their names.

- Added a transfer rate (bytes/sec) indicator to transfer logs.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.21

- Added a fake listing routine to twoftpd-drop, to avoid errors on GUI
  clients that require LIST or NLST.

- Fixed a compile failure on FreeBSD in the newer MDTM command.
  Thanks Sergei Kolobov.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.20

- Modified the backends (twoftpd-auth and twoftpd-xfer) to show a 220
  banner if the front end authenticated, and 230 if not.  This fixes
  problems some FTP clients have when running twoftpd-anon after
  authentication.

- Added twoftpd-drop, a write-only FTP server.

- Added twoftpd-switch, a tool to choose between twoftpd-anon,
  twoftpd-drop, and twoftpd-xfer based on the value of $SHELL,
  $SHELL_READONLY, and $SHELL_WRITEONLY.

- Make sure to open only regular files.

- Added a MDTM command to the xfer backend that can set timestamps
  (instead of just retrieving them).

- Fixed bug exposed by Mozilla which tries to retrieve directories
  before listing them.

- Eliminated the bogus APPE-after-REST test.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.17

- Fixed a bug where paths starting with "/." would not be omitted when
  nodotfiles was set.
- Check the return value from *buf_close functions.
- Fixed a minor (but annoying) typo in listing permissions of sticky
  files.
- Path permission failures (no dot files / lock home) now produce a
  sensible system error.
- Fixed a bug where listing "/foo/*" would not include the "/foo/"
  prefix in the resulting names.
- Updated for bglibs version 1.006

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.16

- Added a port-20 socket hack, based partly on code contributed by
  Wejn.
- Fixed problems with handling interrupted system calls (strace would
  kill twoftpd).
- Converted to use bglibs.
- Added support for restarting uploads with REST.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.15

- Remember to uncork the socket before closing it.

- Fixed a bug in twoftpd-anon-conf -- getpwnam returns the same static
  struct each time it's called, but I used it in two different ways.

- Fixed a minor bug in CWD command to make it possible to cd into
  execute-only directories.

- Fixed handling when doing NLST against non-existant files.

- When making/accepting connections, abort if stdin becomes readable
  (because of additional command or hangup).

- Added support for handling supplemental group IDs.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.14

- Fixed a bug in the string sort implementation that caused listings to
  leak memory.

- Fixed up PASV connection mode to only show the 150 response if the
  connection does not fail.

- Some other minor fixes to socket handling and response messages.

- Added a work around for broken firewalls that rewrite certain commands
  (specifically PORT) with two CR characters before the final LF.

- Fixed missing socket library for SysV (ie Solaris) systems.

- Strip trailing slashes, which confuse at least the "lock $HOME" mode,
  from the home path.

- Use corked (Linux) or no-push (BSD) writes to improve transfer
  efficiency.

- Set SO_LINGER on outbound sockets (LIST/RETR) to prevent the success
  message from being sent on the network before the data is completely
  sent.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.13

- Bugfix: listing "." when CWD is "/" would fail.

- Include system error message in STAT and MDTM.

- Print out an error message when the session or command times out,

twoftpd-xfer.c
/* twoftpd-xfer.c - Main dispatch table for twoftpd-xfer
 * Copyright (C) 2008  Bruce Guenter <bruce@untroubled.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
#include "twoftpd.h"
#include "backend.h"

const char program[] = "twoftpd-xfer";

int store_exclusive = 0;

const command verbs[] = {
  { "TYPE", 0, 0,           handle_type },
  { "STRU", 0, 0,           handle_stru },
  { "MODE", 0, 0,           handle_mode },
  { "CWD",  0, 0,           handle_cwd },
  { "PWD",  0, handle_pwd,  0 },
  { "CDUP", 0, handle_cdup, 0 },
  { "PASV", 0, handle_pasv, 0 },
  { "PORT", 0, 0,           handle_port },
  { "LIST", 0, handle_list, handle_list },
  { "NLST", 0, handle_nlst, handle_nlst },
  { "SIZE", 0, 0,           handle_size },
  { "MDTM", 0, 0,           handle_mdtm2 },
  { "REST", 0, 0,           handle_rest },
  { "RETR", 0, 0,           handle_retr },
  { "STOR", 0, 0,           handle_stor },
  { "APPE", 0, 0,           handle_appe },
  { "MKD",  0, 0,           handle_mkd  },
  { "RMD",  0, 0,           handle_rmd  },
  { "DELE", 0, 0,           handle_dele },
  { "RNFR", 0, 0,           handle_rnfr },
  { "RNTO", 0, 0,           handle_rnto },
  /* Compatibility verbs as defined by RFC1123 */
  { "XCWD", 0, 0,           handle_cwd },
  { "XPWD", 0, handle_pwd,  0 },
  { "XCUP", 0, handle_cdup, 0 },
  { "XMKD", 0, 0,           handle_mkd  },
  { "XRMD", 0, 0,           handle_rmd  },
  /* Handle stray login commands */
  { "USER", 1, 0,           handle_pass },
  { "PASS", 1, 0,           handle_pass },
  { "ACCT", 0, 0,           handle_pass },
  { 0,      0, 0,           0 }
};

const command site_commands[] = {
  { "CHMOD", 0, 0, handle_site_chmod },
  { 0,       0, 0, 0 }
};
